stackit/internal/services/mongodbflex/instance/resource.go

```diff
@@ -0,0 +1,787 @@
+	// Retrieve values from state
+	var model Model
+	diags := req.State.Get(ctx, &model)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	projectId := model.ProjectId.ValueString()
+	instanceId := model.InstanceId.ValueString()
+	ctx = tflog.SetField(ctx, "project_id", projectId)
+	ctx = tflog.SetField(ctx, "instance_id", instanceId)
+
+	// Delete existing instance
+	err := r.client.DeleteInstance(ctx, projectId, instanceId).Execute()
+	if err != nil {
+		core.LogAndAddError(ctx, &resp.Diagnostics, "Error deleting instance", fmt.Sprintf("Calling API: %v", err))
+		return
+	}
+	_, err = wait.DeleteInstanceWaitHandler(ctx, r.client, projectId, instanceId).SetTimeout(15 * time.Minute).WaitWithContext(ctx)
+	if err != nil {
+		core.LogAndAddError(ctx, &resp.Diagnostics, "Error deleting instance", fmt.Sprintf("Instance deletion waiting: %v", err))
+		return
+	}
+	tflog.Info(ctx, "MongoDB Flex instance deleted")
+}
+
+// ImportState imports a resource into the Terraform state on success.
+// The expected format of the resource import identifier is: project_id,instance_id
+func (r *instanceResource) ImportState(ctx context.Context, req resource.ImportStateRequest, resp *resource.ImportStateResponse) {
+	idParts := strings.Split(req.ID, core.Separator)
+
+	if len(idParts) != 2 || idParts[0] == "" || idParts[1] == "" {
+		core.LogAndAddError(ctx, &resp.Diagnostics,
+			"Error importing instance",
+			fmt.Sprintf("Expected import identifier with format: [project_id],[instance_id]  Got: %q", req.ID),
+		)
+		return
+	}
+
+	resp.Diagnostics.Append(resp.State.SetAttribute(ctx, path.Root("project_id"), idParts[0])...)
+	resp.Diagnostics.Append(resp.State.SetAttribute(ctx, path.Root("instance_id"), idParts[1])...)
+	tflog.Info(ctx, "MongoDB Flex instance state imported")
+}
+
+func mapFields(resp *mongodbflex.GetInstanceResponse, model *Model, flavor *flavorModel, storage *storageModel, options *optionsModel) error {
+	if resp == nil {
+		return fmt.Errorf("response input is nil")
+	}
+	if resp.Item == nil {
+		return fmt.Errorf("no instance provided")
+	}
+	if model == nil {
+		return fmt.Errorf("model input is nil")
+	}
+	instance := resp.Item
+
+	var instanceId string
+	if model.InstanceId.ValueString() != "" {
+		instanceId = model.InstanceId.ValueString()
+	} else if instance.Id != nil {
+		instanceId = *instance.Id
+	} else {
+		return fmt.Errorf("instance id not present")
+	}
+
+	var aclList basetypes.ListValue
+	var diags diag.Diagnostics
+	if instance.Acl == nil || instance.Acl.Items == nil {
+		aclList = types.ListNull(types.StringType)
+	} else {
+		acl := []attr.Value{}
+		for _, ip := range *instance.Acl.Items {
+			acl = append(acl, types.StringValue(ip))
+		}
+		aclList, diags = types.ListValue(types.StringType, acl)
+		if diags.HasError() {
+			return fmt.Errorf("failed to map ACL: %w", core.DiagsToError(diags))
```

Comment:
[1] When wrapping errors, don't include "error", "failure" etc. in the message.
The innermost error should already be phrased in such a way that it is clear that something failed. Adding another "failed" every time we wrap it is redundant